### PR TITLE
net: context: Check interface status when sending

### PR DESCRIPTION
**Pull request by jukkar**
If the network interface is down when trying to send a message,
return -ENETDOWN as we cannot send a message in this case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -1571,6 +1570,11 @@ static int context_sendto(struct net_context *context,
 		}
 	}
 
+	iface = net_context_get_iface(context);
```
Comment (chrta) [inline subsys/net/ip/net_context.c:1573]: Should this be moved to the top of this function, like line 1410? In case the interface is down, it would be detected earlier and less unnecessary code is executed.

Comment (jukkar) [reply subsys/net/ip/net_context.c:1573]: The code does various sanitychecks of the destination address etc. so I think we should do those first, and only when all the params are correctly set and we are about to start to allocate the net_pkt, we verify that we can actually send stuff.

**Review by chrta — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by pfalcon — COMMENTED** [review]
This isn't exactly how TCP is supposed to behave, and how other TCP stacks behave. Random glitch in the interface (e.g., WiFi lost carrier and interface went down, just to go up at once again) shouldn't automatically lead to broken TCP connections. So, ideally, this should be configurable. 
Default may be even "on", as such behavior makes sense for common Zephyr usecases. But forcing the stack to be non-compliant without a recurse is a bad idea IMHO.

**Comment by pfalcon** [comment]
"should**n't** automatically lead to broken TCP connections" for those who reads email notifications, I made a typo in the initial comment, fixed it now.

**Comment by pfalcon** [comment]
And the commit talks about the TCP window size - note that its purpose is to *limit* amount of unacked data sent to peer, not to force us to send as much data to it (and buffer it for that). In other words, if peer says its window is 100 bytes, we shouldn't send more than 100 bytes. But if it says it's 1MB, we can still apply a limit on our side, e.g. not buffer, say, more that 2KB (if an app tries to send more, it blocks or gets EAGAIN).

**Review by pfalcon — DISMISSED** [review]
Based on my previous comments, let me give -1, to make sure those points are discussed, instead of prematurely merging this.

**Comment by jukkar** [comment]
@pfalcon I probably will need to rework the commit message as it seems to cause some confusion. Anyway, returning ENETDOWN if the network interface that we are trying to send the message to, is down, is very expected. After all you cannot send a message in this case. It is up to the application to decide what it wants to do in this case, try again later or just quit.

**Comment by pfalcon** [comment]
@jukkar:

> I will rework the commit message as it seems to cause some confusion.

I don't think that's the culprit here, the talk is about the actual semantics of the TCP "send" operation.

> Anyway, returning ENETDOWN if the network interface that we are trying to send the message to, is down, is very expected. After all you cannot send a message in this case.

No, it only means that you cannot send the message *at this instant of time*. TCP is the reliable protocol, and random glitches don't cause catastrophic consequences to a connection. And to remind, the only transient error which may happen with networking operations is EAGAIN (but that should be returned only if enable by application). Any other error is fatal, meaning "this connection is hosed, socket should be closed and if needed, new connection established".

So, let's look again at the time sequence:

1. The interface was up, at we were able connect() via it, or we wouldn't call send() in the first place.
2. We send() successfully.
3. Then the interface glitched and went down.
4. But only for milliseconds, then it went back up.
5. The connection is live, and we continue send()ing via it successfully (perhaps after some packer rexmits).

So again, p.3 shouldn't cause any fatal consequences to an active TCP connection(s), and returning ENETDOWN would make an application think there's (because a random BSD sockets application treats any error but EAGAIN as such).

So, again, specific proposals:

1. If it's *needed now*, make it configurable, to preserve standard BSD Sockets semantics.
2. Longer term, re-check how sending logic works. I still didn't have a chance to dig into TCP2 inner workings. But how it should work is: Any socket should have a send buffer associated with it at the creation. Its size should be changeable via setsockopt(). Only as much data as fits into the send buffer is ever buffered from the app (when it fills, the app is either blocks in send() or gets EAGAIN), so one socket can't starve the other sockets or rest of the system.

**Comment by pfalcon** [comment]
> So again, p.3 shouldn't cause any fatal consequences to an active TCP connection(s), and returning ENETDOWN would make an application think there's (because a random BSD sockets application treats any error but EAGAIN as such).

You can verify this on Linux: telnet to an HTTP server which has human-level connection timeout (seconds, not sub-seconds, sadly many currently are the latter). Type 1st GET line, put your wifi connection down, and then immediately back up, hoping it still reconnects at once. Then type Enter in telnet. You'll still get the response (unless, again, the server times out connection too soon).

**Comment by jukkar** [comment]
> And to remind, the only transient error which may happen with networking operations is EAGAIN (but that should be returned only if enable by application). Any other error is fatal, meaning "this connection is hosed, socket should be closed and if needed, new connection established".

Note that in this case, the issue is not transient but the actual network interface is not functioning (either user turned network interface off or network cable was disconnected). Returning EAGAIN would be quite confusing. Application certainly try again later if it wishes or terminate the connection in this case. The point is that we should provide a errno that is somehow reflecting the current status of the connection.
The ENETDOWN is documented like this in https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_node/libc_17.html
```
Macro: int ENETDOWN
    A socket operation failed because the network was down. 
```

**Comment by pfalcon** [comment]
> Note that in this case, the issue is not transient but the actual network interface is not functioning (either user turned network interface off or network cable was disconnected).

How is that different? User turned network interface off, user turned network interface on. From PoV of TCP, that's not a reason to break an existing connection. And again, any error by EAGAIN means "broken" connection. Simply because EAGAIN is the only error defined as transient, and there can be any number of other errors depending on a particular system. So, portable socket software can rely only on EAGAIN being transient, and that's exactly what the talk is about - to keep the semantics of TCP connections in Zephyr generic and not introduce unneeded behaviorial discrepancies.

> Returning EAGAIN would be quite confusing.

EAGAIN can be returned *only* if app explicitly sanctioned that (made socket non-blocking or set send timeout on it), and then it's not confusing at all, it's exactly what an app expects.

> Better reference here https://pubs.opengroup.org/onlinepubs/009604599/functions/send.html

It says that it *may* return it. And I tell you that Linux doesn't return it (under normal "interface went down" conditions). And Linux handling follows the TCP semantics, and that's what I propose to do either.

So, let me try again: Would you please make the changes in this patch Kconfig'uable (y/n)?

**Comment by pfalcon** [comment]
Anyway, I don't have other arguments, so removing my -1, do as you see fit.
